src/net_processing.cpp

```diff
@@ -991,6 +991,39 @@ void static ProcessGetData(CNode* pfrom, CConnman& connman, std::atomic<bool>& i
     }
 }
 
+static void CheckBlockSpam(CValidationState& state, CNode* pfrom, const uint256& hashBlock)
+{
+    // Block spam filtering
+    if (!GetBoolArg("-blockspamfilter", DEFAULT_BLOCK_SPAM_FILTER)) {
+        return;
+    }
+    CNodeState *nodestate = State(pfrom->GetId());
+    if(!nodestate) {
+        return;
+    }
+
+    const auto it = mapBlockIndex.find(hashBlock);
+    if (it == mapBlockIndex.end()) {
```

Comment:
This is a redundant check.

Before calling `ProcessNewBlock` we are checking if the block is already on `mapBlockIndex` or not.

Reply:
Yes, **before**.
But that should change **after**: `ProcessNewBlock` calls  `AcceptBlock` which adds the block to the index (via `AcceptBlockHeader`), so this is not redundant.

Reply:
yeah ok, missed that flow.

So, the rationality behind this is that because of the block processing failed on the header check, it doesn't represent a spam risk.

Reply:
yes, exactly.